### PR TITLE
[MLIR] Add padding support to Convolution

### PR DESCRIPTION
**Pull request by nmostafa**
Do a range check on the Img spatial indices and perform computation if within non-padded region. 

Depends on #3584

### CODE REVIEW

File: src/contrib/mlir/compiler/lowerer.cpp [hunk 1]
```diff
@@ -804,11 +794,13 @@ namespace
         //           Output[n, k, r_1, .. r_f] +=
         //             Images[n, c, i_1 + j_1, .. i_f + j_f] * Filters[k, c, j_1, .. j_f]
 
-        // TODO: With padding, we need to check (using IntegerSets) whether each spatial dim in
-        // Images lie within paddings
-        // If yes, we init value to zero, else load from MemRef.
-        // Q: Can this be done using a map from padded tensor to  unpadded one ? Will we load zero
-        // if OOB ?
+        // With padding, we check (using IntegerSets) whether each spatial dim in Images lie within
+        // paddings. If true, we perform the computation:
```
Comment (dcaballe) [inline src/contrib/mlir/compiler/lowerer.cpp]: "lie within paddings" sounds as if we performed computation for elements in the padding. Should we say outside of padding instead?

File: src/contrib/mlir/compiler/lowerer.cpp [hunk 2]
```diff
@@ -842,6 +837,22 @@ namespace
...
+                withPadding = true;
+            }
+            padBelowIntValues.push_back(padValue);
+
+            iAttr = padAbove[i].cast<IntegerAttr>();
+            padValue = iAttr.getInt();
+            if (padValue)
+            {
+                withPadding = true;
+            }
```
Comment (dcaballe) [inline src/contrib/mlir/compiler/lowerer.cpp]: For my understanding... we don't need `padAboveValues`?

File: src/contrib/mlir/compiler/lowerer.cpp [hunk 3]
```diff
@@ -860,6 +871,41 @@ namespace
...
+        IntegerSet nonPaddedRange;
+        if (withPadding)
+        {
+            // Create affine expressions and IntegerSet
+            // IntegerSet (d0, d1, .. d_N-1)[LB_0, LB_1, .. LB_N-1, UB_0, UB_1, .. UB_N-1], where
+            // for each dim:
+            //   (d_dim - padBelow[dim] - LB_dim >= 0),
+            //   (padBelow[dim] + UB_dim - d_dim - 1 >= 0)
+            SmallVector<AffineExpr, 4> affineExprs;
+            SmallVector<bool, 4> isEq;
```
Comment (dcaballe) [inline src/contrib/mlir/compiler/lowerer.cpp]: Please, doc for `isEq`

File: src/contrib/mlir/compiler/lowerer.cpp [hunk 4]
```diff
@@ -928,8 +976,47 @@ namespace
                                                   filtersSpatialIndices.begin(),
                                                   filtersSpatialIndices.end());
 
-                            iRes(resIndices) =
-                                iRes(resIndices) + (iImages(imgIndices) * iFilters(filtersIndices));
+                            if (withPadding)
+                            {
+                                // if args : img dims, img lbs, img ubs
+                                SmallVector<IndexHandle, 4>::iterator it = imgIndices.begin();
```
Comment (dcaballe) [inline src/contrib/mlir/compiler/lowerer.cpp]: std::advance(imgIndices.begin(), 2)

File: test/backend/convolution.in.cpp [hunk 5]
```diff
@@ -99,3 +99,40 @@ NGRAPH_TEST(${BACKEND_NAME}, convolution_simple)
...
+    vector<float> expected_result{0.0f, 0.0f, 0.0f, 0.0f, 0.0f,
+                                  0.0f, 2.0f, 4.0f, 0.0f, 0.0f,
+                                  0.0f, 6.0f, 8.0f, 0.0f, 0.0f,
+                                  0.0f, 0.0f, 0.0f, 0.0f, 0.0f,
+                                  0.0f, 0.0f, 0.0f, 0.0f, 0.0f};
+    // clang-format on
+    auto handle = backend->compile(f);
+    handle->call_with_validate({result}, {a, b});
+    EXPECT_TRUE(test::all_close_f(vector<float>{expected_result}, read_vector<float>(result)));
+}
```
Comment (dcaballe) [inline test/backend/convolution.in.cpp]: Could you please add a lit test to cover the lowering of a convolution with padding? For example to `mlir/lowering/convolution.mlir`

Comment (nmostafa) [reply test/backend/convolution.in.cpp]: Wouldn't we still need to fix ngraph-opt to run the lowering pass ?

File: src/contrib/mlir/compiler/lowerer.cpp (hunk 2 above)
Comment (nmostafa) [reply src/contrib/mlir/compiler/lowerer.cpp]: Not to the check if we are inside the non-padded range. PadAbove is actually included in the inferred output shape early on in nGraph, so the loop bounds are adjusted here to include padAbove though.

File: test/backend/convolution.in.cpp (hunk 5 above)
Comment (dcaballe) [reply test/backend/convolution.in.cpp]: Yes, working on that. Sorry

**Pull request merged**